readme.txt
UOLog Beta 1 Readme

1 What is UOLog?
UOLog is a packet logging tool for Ultima Online.

2 Am I allowed to use UOLog?
You may only use UOLog if you are a shard's administrator or if you have his permission.
You are _not_ allowed to sniff packets on OSI, it's a violation of their Terms of Use (TOS).

3 Why can't I use Attach in Windows 98?
I need the handle of the client's thread. In Windows 98 I can only get it if I am the creator
of the process. In newer Windows versions I can use OpenThread to get the thread handle.
I know there is a way to get the threadhandle in Windows98, if you know it, please tell me.

Quote from MSDN:
"There is no way to get the thread handle from the thread ID. While there is an 
OpenProcess() API that takes a PID and returns the handle to the process, there 
is no corresponding OpenThread() that takes a thread ID and returns a thread 
handle. The reason that the Win32 API does not make thread handles available 
this way is that it can cause damage to an application. The APIs that take a 
thread handle allow suspending/resuming threads, adjusting priority of a thread 
relative to its process, reading/writing registers, limiting a thread to a set 
of processors, terminating a thread, and so forth. Performing any one of these 
operations on a thread without the knowledge of the owning process is dangerous, 
and may cause the process to fail." Source: Microsoft Developer Network, Q127992

However, in Windows 2000 they added OpenThread. No further comments.

4 How do I find the values for clients.cfg?
Check finding_values.html.

5 Sourecode?
Nay, not until it's out of beta.

6 Other
Post in my forums (http://uo.elitecoder.net/phpBB/).

7 Credits
Arog <arog@elitecoder.net> for autoscrolling.
DarkStorm for the idea of using EXCEPTION_SINGLE_STEP instead of the DeBPXs.

--
Folko, 2003, somewhere near Hamburg, Germany.


changes.txt
Note: The "Qxxxxxx" are Microsoft Support IDs. You can look up the documents on http://www.microsoft.com

This is the last UOLog 1.x version.
The next release will be completely rewritten, featuring a plugin engine and other things.

Release 1.2:

17 April 2004:
    -Added a new filter type: HexLex
     It is a regular expression engine which I coded for UOLog.
     You can enter filter strings like "BF ?? ?? ?? 06|07" to only log BF SubType 06 or 07.
     The syntax is easy, seperate each byte with a space. '|' is or. '?' is a wildcard for any nibble.
     ?? can be used as wildcard on any byte.
     Example:
     "02 ?2|?3" would only log walking requests for direction 02 or 03, running and walking.
     Note that invalid strings (spaces before or after | etc) might crash, I didn't want to debug it yet.
     Have fun with this new feature.

Release 1.1:

27 August 2003:
   -Added Logging -> Raw, handy if you want to paste the output directly into a script.
   -Fixed some minor typos in the tutorial.


Release 1.0:

03 August 2003:
   -Copying from the logwindow is not in unix format now.
   -The millisecond part of the timestamp is correct now.


Beta 9:

10 June 2003:
   -Fixed a bug where detaching would crash the client.
   -Compiled with Visual Studio .NET 2003, hope that doesn't break anything *ducks*.

Beta 8:

7 June 2003:
   -Q121093: Fixed a bug in the DebugLoop which made UOLog swallow a lot of packets. Now it's even working on RunUO.
   -Q108315: Added "Always on Top" option.
   -Q109550: Changed the PacketWindow's internal behaviour. It won't need memory anymore, that means it can hold
    an unlimited amount (Well, 4Gb) of text without lagging like older versions.
    I had 1Mb of text in the logwindow and it didn't lag, before it started to lag at 100kb.
    Unfortunately this broke AutoScrolling a bit, if you have the scrollbar at the bottom of the window,
    it will still autoscroll, regardless of the settings. But if you move the bar up one pixel, it won't scroll.
    I think that's better than a lagging logwindow.
   -Assertions now generate dumps, too. UOLog still didn't crash for me yet, after 6 months of using it.
   -Converted the whole project to pure C.
   -Added more packet names.

Beta 7:

5 June 2003:
   -*If* UOLog crashes, it will now generate CrashDumps (like POL's).
    Two crashdumps will be generated, a small and a big one.
    This feature requires DbgHelp.dll, included in Windows XP, versions
    for older Windows' are downloadble from Microsoft.com.
    If you can compress the big one to <=3Mb, please mail it to me.
    Else mail me the other one.
    Note that the big one might contain your account data (The big one
    sends UOLog's memory, which contains the LogWindow contents, which
    may contain your login packets).

   -Fixed a very stupid error which was introduced in Beta 5.
    The error was that the filter didn't work, resulting in no packets
    being logged. Heh.

Beta 6:

2 June 2003:
   -DeBPXs are no longer required! Thanks to DarkStorm for the idea with EXCEPTION_SINGLE_STEP.
   -Options are no longer saved automatically, they are saveable via Options->SaveOptions and
    automatically loaded at startup.

Beta 5:

8 Feb 2003:
   -Added automatic option saving and loading.

6 Feb 2003:
   -Internal code cleanup to prepare the open source release.

Beta 4:

4 Feb 2003:
   -Added Autoscrolling. Thanks to Arog <arog@elitecoder.net> for this feature.
   -Fixed a bug when UO runs on a different CPU (on dual CPU systems) than UOLog.
   -Fixed a major bug where UOLog would print random stuff into the packetlog if the
    packet contained a '%' char (Reason was using fprintf for printing to the log; now using
    fwrite).
   -The packet statistics are now printed into the log even if you close UOLog before closing the
    client.

Beta 3:

15 Jan 2003:
   -Added automatic log flushing.
   -Optimized GetPacketName.
   -Fixed log clearing.

Beta 2:

14 Jan 2003:
   -Fixed a bug in the BPX handling which made UOLog unable to use on Win98 / 2000.
   -Released.

10 Jan 2003:
   -Changed old 'logwindow' to a richedit wrapperclass. Still in development.

09 Jan 2003:
   -Improved my Breakpoint wrapperclass to be faster.

08 Jan 2003:
   -Major code cleanup, reduced size from 1200 lines of code to 800 lines of code.
   -Improved the performance.

07 Jan 2003:
   -Minor code cleanups. Should run more stable now (It never crashed for me yet).

06 Jan 2003:
   -Fixed a very small resource leak (Clienthandle not being CloseHandle()'d).
   -Added assertions to help detecting bugs.

05 Jan 2003:
   -Added validation of RecvBuffer and SendBuffer.
   -If you specify 32 bit registers for length,
    UOLog will use the 16 bit ones now. That is due to
    changes in newer clients. (Internally: 32 bit reg AND 0xFFFF)

Beta 1:

04 Jan 2003:
   -Initial release.


Debugger.c
/**************************************************************************************\
*                                                                                      *
*  This file is part of UOLog,                                                         *
*  An advanced packet monitoring and logging tool for Ultima Online (UO).              *
*                                                                                      *
*  Copyright (C) 2003 by Folke Will (Folko)                                            *
*                                                                                      *
*  This program is free software; you can redistribute it and/or modify                *
*  it under the terms of the GNU General Public License as published by                *
*  the Free Software Foundation; either version 2 of the License, or                   *
*  (at your option) any later version.                                                 *
*                                                                                      *
*  This program is distributed in the hope that it will be useful,                     *
*  but WITHOUT ANY WARRANTY; without even the implied warranty of                      *
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                       *
*  GNU General Public License for more details.                                        *
*                                                                                      *
*  You can contact the author by sending E-Mail to Folko <Folko@elitecoder.net>        *
*                                                                                      *
\**************************************************************************************/

/*
    <CVS SERVER GENERATED>
    This project was stored on Folko's CVS server as $Source: /home/cvs/UOLogC/Debugger.c,v $.
   
    Id tag follows (<filename> <revision> <last changed> <author> <status>):
    $Id: Debugger.c,v 1.3 2003/08/04 02:09:58 folko23 Exp $
   
    If you want to report any errors, please include this tag.
    </GENERATED>
*/

#include "UOLog.h"

int detach;

void DebugLoop(void *param)
{
   short keepalive=1;
   int attached;
   DEBUG_EVENT dbgev;
   DWORD TargetPID=((DebugLoopParam*)param)->TargetPID;
   HANDLE TargetProcess=0, Thread=0;
   PROCESS_INFORMATION info;
   Breakpoint BPXSend, BPXRecv;
   
   detach=0;
   attached=((DebugLoopParam*)param)->attached;
   if(attached)
      TargetProcess=OpenProcess(PROCESS_ALL_ACCESS, 0, TargetPID);
   else
   {
      char *path=((DebugLoopParam*)param)->path;
      STARTUPINFO start;
      int res=0;

      memset(&info, 0, sizeof(info));
      memset(&start, 0, sizeof(start));
      start.cb=sizeof(start);
      res=CreateProcess(path, NULL, NULL, NULL, 0, CREATE_DEFAULT_ERROR_MODE | NORMAL_PRIORITY_CLASS, 0, 0, &start, &info);
      if(res)
      {
         Thread=info.hThread;
         TargetProcess=info.hProcess;
         TargetPID=info.dwProcessId;
         check(Thread!=0);
         WaitForInputIdle(TargetProcess, INFINITE);
      }
   }
   check(TargetPID!=0);
   check(TargetProcess!=0);
   if(!DebugActiveProcess(TargetPID))
   {
      MessageBox(Options.hWnd, "Couldn't get debug access to UO. Make sure the exe is readable and not opened by anything else.", "Error", MB_ICONERROR);
      EnableMenuItem(Options.MainMenu, ID_CLIENT_ATTACH, MF_ENABLED);
      EnableMenuItem(Options.MainMenu, ID_CLIENT_DETACH, MF_GRAYED);
      EnableMenuItem(Options.MainMenu, ID_CLIENT_LOAD, MF_ENABLED);
      _endthread();
      return;
   }
   while(keepalive)
   {
      DWORD status=DBG_CONTINUE;

      WaitForDebugEvent(&dbgev, INFINITE);
      switch(dbgev.dwDebugEventCode)
      {
         case CREATE_PROCESS_DEBUG_EVENT: /* This is caused after we attached successfully, so let's set the breakpoints here. */
            {
               char *res=SetupDebugger(&dbgev, &Client);
               if(res)
               {
                  MessageBox(Options.hWnd, res, "Fatal error", MB_ICONERROR);
                  keepalive=0;
                  break;
               }
               BPXSend.Address=Client.Send;
               BPXSend.CallbackFunc=OnPacket;
               BPXSend.dbgev=&dbgev;
               BPXSend.TargetProcess=TargetProcess;
               BPXSend.type=TYPE_SEND;
               SetBreakpoint(&BPXSend);

               BPXRecv.Address=Client.Recv;
               BPXRecv.CallbackFunc=OnPacket;
               BPXRecv.dbgev=&dbgev;
               BPXRecv.TargetProcess=TargetProcess;
               BPXRecv.type=TYPE_RECV;
               SetBreakpoint(&BPXRecv);

               if(Options.FileLogging)
                  res=InitLogger();
               else
                  res=0;
               if(res)
               {
                  MessageBox(Options.hWnd, res, "Error", MB_ICONERROR);
                  keepalive=0;
                  break;
               }
               AppendLogWindow(Options.PacketLog, TimeStamp("Detected client %s\n", Client.Version));
               status=DBG_CONTINUE;
            }break;
         case EXCEPTION_DEBUG_EVENT:
            {
               if(dbgev.u.Exception.ExceptionRecord.ExceptionCode==EXCEPTION_BREAKPOINT) /* BPX */
               {
                  void *addr=dbgev.u.Exception.ExceptionRecord.ExceptionAddress;;
                  if(attached)
                  {
                     typedef HANDLE (WINAPI *_Open)(DWORD, int, DWORD);
                     HINSTANCE kernel=LoadLibrary("KERNEL32");
                     _Open Open=(_Open)GetProcAddress(kernel, "OpenThread");
                     Thread=Open(THREAD_ALL_ACCESS, 0, dbgev.dwThreadId);
                     FreeLibrary(kernel);
                  }
                  if(addr==BPXSend.Address && !detach){ /* BPXSend caught. Dump the packet and set SINGLE_STEP (via Action). */
                     BreakpointAction(&BPXSend, Thread);
                     status=DBG_CONTINUE;
                  }else if(addr==BPXRecv.Address && !detach){
                     BreakpointAction(&BPXRecv, Thread);
                     status=DBG_CONTINUE;
                  }
                  if(attached)
                     CloseHandle(Thread);
               }else if(dbgev.u.Exception.ExceptionRecord.ExceptionCode==EXCEPTION_SINGLE_STEP){
                     /* Process is in single step mode. Means we have to re-set our breakpoints. */
                     SetBreakpoint(&BPXSend);
                     SetBreakpoint(&BPXRecv);
                     status=DBG_CONTINUE;
               }else{ /* Process caused exception, so let it handle the exception! */
                  status=DBG_EXCEPTION_NOT_HANDLED;
               }
            }break;
         case EXIT_PROCESS_DEBUG_EVENT: /* Client closed */
            {
               AppendLogWindow(Options.PacketLog, TimeStamp("Client terminated\n"));
               keepalive=0;
               status=DBG_CONTINUE; 
            }break;
         default:
            status=DBG_CONTINUE;
      }
      if(detach) /* User wants to detach the debugger. We have to remove all BPXs first. */
      {
         RemoveBreakpoint(&BPXSend);
         RemoveBreakpoint(&BPXRecv);
         keepalive=0;
      }
      ContinueDebugEvent(dbgev.dwProcessId, dbgev.dwThreadId, status);
   }
   if(detach)
   {
      typedef int (WINAPI *_Stop)(DWORD);
      HINSTANCE kernel=LoadLibrary("KERNEL32");

      _Stop Stop=(_Stop)GetProcAddress(kernel, "DebugActiveProcessStop");
      Stop(TargetPID);
      FreeLibrary(kernel);
      AppendLogWindow(Options.PacketLog, TimeStamp("Detached successfully\n"));
   }
   if(!attached)
      CloseHandle(Thread);
   CloseHandle(TargetProcess);
   FinalizePacketLog();
   EnableMenuItem(Options.MainMenu, ID_CLIENT_ATTACH, MF_ENABLED);
   EnableMenuItem(Options.MainMenu, ID_CLIENT_DETACH, MF_GRAYED);
   EnableMenuItem(Options.MainMenu, ID_CLIENT_LOAD, MF_ENABLED);
   EnableMenuItem(Options.MainMenu, ID_LOGGING_LOGFILE, MF_ENABLED);
   _endthread();
}

unsigned long GetTimeStamp(HANDLE file)
{
   unsigned long ClientSize;
   unsigned char *image;
   IMAGE_DOS_HEADER *doshdr;
   IMAGE_FILE_HEADER *filehdr;
   DWORD temp;

   ClientSize=GetFileSize(file, 0);
   image=(unsigned char *)malloc(sizeof(unsigned char)*ClientSize);
   if(!image) return 0;
   if(!ReadFile(file, image, ClientSize, &temp, 0)) {free(image); return 0;}
   doshdr=(IMAGE_DOS_HEADER*)image;
   filehdr=(IMAGE_FILE_HEADER*)(image+doshdr->e_lfanew+sizeof(IMAGE_NT_SIGNATURE));
   temp=filehdr->TimeDateStamp;
   free(image);
   return temp;
}

char *SetupDebugger(DEBUG_EVENT *dbgev, ClientEntry *Client)
{
   FILE *cfgfile;
   char line[MAX_LINE], timestring[10];
   int res, found=0;
   DWORD stamp=0;

   if(!dbgev->u.CreateProcessInfo.hFile) return "Can't get access to client executable..";
   cfgfile=fopen("clients.cfg", "r"); if(!cfgfile) return "Can't open clients.cfg..";

   stamp=GetTimeStamp(dbgev->u.CreateProcessInfo.hFile);
   if(!stamp)
   {
      fclose(cfgfile);
      return "Couldn't get the client's timestamp..";
   }
   sprintf(timestring, "%X", stamp);
   while(fgets(line, sizeof(line), cfgfile))
   {
      char *offset;
      if(strstr(line, timestring))
      {
         int j=0, i=0;
         found=1;
         offset=strchr(line, '\"');
         if(!offset)
         {
            fclose(cfgfile);
            return "Missing an opening \" in clients.cfg";
         }
         if(!strchr(offset+1, '\"'))
         {
            fclose(cfgfile);
            return "Missing a closing \" in clients.cfg";
         }
         while(*(offset+(++i))!='\"') Client->Version[j++]=offset[i];
         Client->Version[i-1]='\0';
         i++;

         res=sscanf(offset+i, "%X %d %d %X %d %d",
            &Client->Send,
            &Client->RegBufSend,
            &Client->RegLenSend,
            &Client->Recv,
            &Client->RegBufRecv,
            &Client->RegLenRecv
            );
      }
   }
   fclose(cfgfile);
   if(!found)
      return "Couldn't find client in clients.cfg";
   if(res!=6)
      return "Clients.cfg seems to have an invalid format";

   check(Client->Send!=0);
   check(Client->RegBufSend!=0);
   check(Client->RegLenSend!=0);

   check(Client->Recv!=0);
   check(Client->RegBufRecv!=0);
   check(Client->RegLenRecv!=0);
   return 0;
}

void Detach(void)
{
   AppendLogWindow(Options.PacketLog, TimeStamp("The next event will detach UOLog.\n"));
   detach=1;
   return;
}

unsigned long GetRegisterContent(CONTEXT *context, short registernum)
{
   check((registernum>0 && registernum<8));
   switch(registernum)
   {
      case 1: return context->Eax; break;
      case 2: return context->Ebx; break;
      case 3: return context->Ecx; break;
      case 4: return context->Edx; break;
      case 5: return context->Esi; break;
      case 6: return context->Edi; break;
      case 7: return context->Ebp; break;
   }
   return 0;
}

